### ISSUE TITLE
Flow generic class members

### ISSUE THREAD

**Issue by leebyron** [babel/babel-eslint#123]
babel-eslint currently produces "unused" errors for valid flow-typed javascript when using generics with class members:

```js
function identity<T>(value: T): T {
  return value;
}
console.log(identity('foo'));

class Box<T> {
  value: T;
  constructor(value: T) {
    this.value = T;
  }
}
var box = new Box();
console.log(box.value);
```

Most uses of the generic seem to be fine, but typing the class members fails lint.

```
   7:9  error  "T" is not defined             no-undef
```

**Comment by hzoo** [comment]
I guess that last patch didn't cut it haha.

Kind of unrelated but is the feature called "generic types?" https://github.com/babel/babel-eslint/pull/109#issuecomment-107781203 used "polymorphic types" and that's what the flow docs were saying http://flowtype.org/docs/functions.html#polymorphic-functions. Generics makes sense too but I was just wondering.

**Comment by leebyron** [comment]
Ah yeah, Flow uses "polymorphic types" to describe this. The concepts are more or less interchangeable in this context though :)

**Comment by hzoo** [comment]
Ah so the reason is because I'm just ignoring/not visiting the those types.
Seems like 3 types to account for: built-in types like `number, boolean, any, mixed, etc`, user defined/imported types like `x: Foo`, and now these polymorphic ones.

Looks like we'll need to create a variable in the scope whenever there is a polymorphic type to fix this.

**Comment by leebyron** [comment]
That sounds like a great way to handle this. function and class are the places polymorphic types can be introduced, you could just create a new scope for each and supply the polymorphic types in that scope

**Comment by sebmck** [comment]
Might be worth looking into shimming `escope` with Babel's scope tracker.

**Comment by hzoo** [comment]
@leebyron Kind of doing something like that now for type definitions https://github.com/babel/babel-eslint/blob/master/index.js#L249-L263
as well as in comprehensions

```js
  function createScopeVariable (node, name) {
    this.currentScope().variableScope.__define(name,
      new Definition(
        "Variable",
        name,
        node,
        null,
        null,
        null
      )
    );
  }

  referencer.prototype.TypeAlias = function(node) {
    createScopeVariable.call(this, node, node.id);
   ...
}
```

@sebmck Ok but will probably need some guidance :smile: Looks like it's https://github.com/babel/babel/tree/master/src/babel/traversal/scope?

**Comment by sebmck** [comment]
@hzoo 

> Ok but will probably need some guidance :smile: 

Was just an idea :smile: It's likely non-trivial (`escope` has a bizzare format and architecture) so don't feel pressured.
